### PR TITLE
fix(kuma-cp): concurrent mesh cache map write

### LINKED ISSUE #4282: Cleanup mesh contexts after Mesh deletion
We should cleanup the [`latestMeshContext` map](https://github.com/kumahq/kuma/blob/master/pkg/xds/cache/mesh/cache.go) when `Mesh`es are deleted.

> Very nice find!

> The only problem that I can see with it is that we never clear mesh contexts from the map
> This is a potential "memory leak", but only with the following scenario
> 1) Create a new mesh
> 2) Connect a DP (or many of them)
> 3) Delete a mesh
> do those steps in the loop.

> Should we write a cleanup mechanism for it?

_Originally posted by @jakubdyszkiewicz in https://github.com/kumahq/kuma/pull/4267#pullrequestreview-968985839_

### LINKED ISSUE #5245: kuma-cp concurrent map write
### What happened?

Looks to be at https://github.com/kumahq/kuma/blob/9b2c75323c59daaa0f755dc7b1d7563375cd5fee/pkg/xds/cache/mesh/cache.go#L47

```
fatal error: concurrent map writes

goroutine 917 [running]:
runtime.throw({0x28308fe?, 0x256e840?})
	/home/mike/.local/share/asdf/installs/golang/1.18.5/go/src/runtime/panic.go:992 +0x71 fp=0xc001653608 sp=0xc0016535d8 pc=0x4385d1
runtime.mapassign_faststr(0x235d620, 0xc000e00d20, {0xc00149a690, 0x7})
	/home/mike/.local/share/asdf/installs/golang/1.18.5/go/src/runtime/map_faststr.go:295 +0x38b fp=0xc001653670 sp=0xc001653608 pc=0x4135eb
github.com/kumahq/kuma/pkg/xds/cache/mesh.(*Cache).GetMeshContext.func1({0x346c3f0, 0xc001415bc0}, {0xc0014867c0?, 0x0?})
	/home/mike/projects/kuma/pkg/xds/cache/mesh/cache.go:47 +0x33a fp=0xc001653800 sp=0xc001653670 pc=0x1507d1a
github.com/kumahq/kuma/pkg/xds/cache/once.RetrieverFunc.Call(0xc00000ee30?, {0x346c3f0?, 0xc001415bc0?}, {0xc00149a690?, 0xc00149ede0?})
```

### PR DESCRIPTION
**Pull request by michaelbeaumont**
Advantage over https://github.com/kumahq/kuma/pull/5252 and the alternative there being that locking is handled by the data structure, the `expirationTime` time check is handled by the data structure and this eventually cleans up unused meshes.

Closes #5245 
Closes #4282 

### Checklist prior to review

- [ ] Link to docs PR or issue --
- [ ] Link to UI issue or PR --
- [ ] Is the [issue worked on linked][1]? --
- [ ] The PR does not hardcode values that might break projects that depend on kuma (e.g. "kumahq" as a image registry) --
- [ ] The PR will work for both Linux and Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [ ] Unit Tests --
- [ ] E2E Tests --
- [ ] Manual Universal Tests --
- [ ] Manual Kubernetes Tests --
- [ ] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [ ] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? --
- [ ] Do you need to explicitly set a [`> Changelog:` entry here](../blob/master/CONTRIBUTING.md#submitting-a-patch) or add a `ci/` label to run fewer/more tests?

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

File: pkg/xds/cache/mesh/cache.go [hunk 1]
```diff
@@ -15,12 +16,21 @@ import (
...
+	// hashCache keeps a cached context, for a much longer time, that is only reused
+	// when the mesh hasn't changed.
+	hashCache *cache.Cache
 
-	latestMeshContext map[string]*xds_context.MeshContext
+	meshContextBuilder xds_context.MeshContextBuilder
 }
 
+// recalculationTime is the time after which the mesh context is recalculated no
+// matter what.
```
Comment (jakubdyszkiewicz) [inline pkg/xds/cache/mesh/cache.go:30]: Why do we need to recalculate no matter what?
Wouldn't it be better to "extend the longer cache lease" every time we compute that the mesh context did not change?

So the logic will be to always call `c.hashCache.SetDefault(mesh, meshCtx)` if we don't have an error after `c.meshContextBuilder.BuildIfChanged(ctx, mesh, context.Hash)`. This way we could
1) Change the expiration time of a longer cache to something shorter (like 1 minute?) causing deleted mesh contexts to be cleaned up faster
2) Avoid recomputing mesh context every hour if nothing changes

Comment (michaelbeaumont) [reply pkg/xds/cache/mesh/cache.go:30]: Yeah that makes sense

**Review by jakubdyszkiewicz — APPROVED** [review]

**Comment by bartsmykla** [comment]
@michaelbeaumont could we merge this PR?

**Comment by michaelbeaumont** [comment]
Maybe @lobkovilya can take a look again too?

**Review by lobkovilya — APPROVED** [review]

**Pull request merged**